src/hp/derp/client.rs

```diff
@@ -715,3 +727,130 @@ pub(crate) fn parse_recv_frame(frame: &[u8]) -> Result<(PublicKey, &[u8])> {
         &frame[PUBLIC_KEY_LENGTH..],
     ))
 }
+
+/// Combines blobs into packets of at most MAX_PACKET_SIZE.
+///
+/// Each item in a packet has a little-endian 2-byte length prefix.
+pub struct PacketizeIter<I: Iterator> {
+    iter: std::iter::Peekable<I>,
+    buffer: BytesMut,
+}
+
+impl<I: Iterator> PacketizeIter<I> {
+    /// Create a new new PacketizeIter from something that can be turned into an
+    /// iterator of slices, like a Vec<Bytes>.
+    pub fn new(iter: impl IntoIterator<IntoIter = I>) -> Self {
+        Self {
+            iter: iter.into_iter().peekable(),
+            buffer: BytesMut::with_capacity(MAX_PACKET_SIZE),
+        }
+    }
+}
+
+impl<I: Iterator> Iterator for PacketizeIter<I>
+where
+    I::Item: AsRef<[u8]>,
+{
+    type Item = Bytes;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        use bytes::BufMut;
+        while let Some(next_bytes) = self.iter.peek() {
+            let next_bytes = next_bytes.as_ref();
+            let next_length: u16 = next_bytes.len().try_into().expect("items < 64k size");
+            if self.buffer.len() + next_length as usize + 2 > MAX_PACKET_SIZE {
+                break;
+            }
+            self.buffer.put_u16_le(next_length);
+            self.buffer.put_slice(next_bytes);
```

Comment:
this means we have to do a full copy for every packet, why can't we just write directly to the io instead?

Reply:
The way to write io seems to be write_frame. It takes a &[&[u8]]. If I understand it correctly it will concatenate all the slices. Since the number of transmits per packet is not known before, you would have to allocate something in any case.

I guess you could rearrange it so that thing is smaller than the buffer, but at that point I doubt it would make a difference.

If you wanted to you could just have this format:

```
2 byte number of transmits
2 * n bytes sizes for each transmit
rest concatenated transmits
```

But I can't see how you can do this entirely without allocations given that you don't have direct access to the AsyncWrite.

Reply:
OKOK, you could do it like this:

```rust
    let mut i = 0;
    for (size, count) in PacketizeIter2::new(transmits.iter()) {
        write_frame_header(&mut writer, FrameType::SendPacket, size).await?;
        for transmit in &transmits[i..i + count] {
            let transmit = transmit.as_ref();
            writer.write_u16_le(transmit.len() as u16).await?;
            writer.write_all(transmit).await?;
        }
        i += count;
    }
```

This has an iterator that just computes the number of transmits that fit into a packet and the total size, then manually write the header and data.

This does not allocate, but it is probably *slower* because it interacts with the writer more often, whereas the packetizing code above just has 1 big call to write_all per packet.

🤷‍♂️

Reply:
so the fastest should be a bufwrite wrapper with your latest version I suspect

Reply:
but lets leave it as is for now, we can optimize things later, when we see bottlenecks